src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -218,6 +220,50 @@ public UnfilteredRowIterator executePartitionReadCommand(SinglePartitionReadComm
         return command.queryMemtableAndDisk(cfs, executionController);
     }
 
+    public RangeIterator buildIterator()
+    {
+        var filterOperation = filterOperation();
+        var orderings = filterOperation.expressions()
+                                       .stream().filter(e -> e.operator() == Operator.ANN).collect(Collectors.toList());
+        assert orderings.size() <= 1;
+        if (filterOperation.expressions().size() == 1 && filterOperation.children().isEmpty() && orderings.size() == 1)
+            // If we only have one expression, we just use the ANN index to order and limit.
+            return getTopKRows(orderings.get(0));
+
+        // We already decided we need to do first sort then filter, so no need to open the index iterator:
+        if (queryContext.filterSortOrder() == QueryContext.FilterSortOrder.SORT_THEN_FILTER)
+        {
+            assert queryContext.postFilterSelectivityEstimate() != null;
+            return getTopKRows(orderings.get(0));
+        }
```

Comment:
When will we hit this condition?

Reply:
On the second iteration of the shadow loop.